File: Backend/models/tenant.py
from datetime import datetime
from enum import Enum
from html import escape as html_escape
from typing import TYPE_CHECKING, Optional, Any
from uuid import UUID as PythonUUID, uuid4
import re

from sqlalchemy import (Column, DateTime, ForeignKey, Index, Integer, String,
                        text)
from sqlalchemy.dialects.postgresql import UUID as PG_UUID, JSONB
from sqlalchemy.sql import func
from sqlmodel import Field, Relationship, SQLModel, col
from pydantic import field_validator

from Backend.utils.datetime_utils import create_audit_datetime

# Use TYPE_CHECKING to avoid circular imports at runtime
if TYPE_CHECKING:
    from Backend.models.lease import Lease
    from Backend.models.property import Property
    from Backend.models.units import PropertyUnit
    from Backend.models.user import User
    from Backend.models.maintenance import MaintenanceRequest
    from Backend.models.accounting.payment import Payment
    from Backend.models.accounting.invoice import Invoice


class TenantStatus(str, Enum):
    ACTIVE = "Active"
    INACTIVE = "Inactive"
    PENDING = "Pending"
    EVICTED = "Evicted"
    MOVED_OUT = "Moved Out"


# Import TenantType from enums
from Backend.models.enums import TenantType


# Link table for tenant-unit many-to-many relationship


class TenantUnitLink(SQLModel, table=True):
    __tablename__ = "tenant_unit_link"  # type: ignore
    __table_args__ = (Index("ix_tenant_unit_link_unit_id", "unit_id"),)

    tenant_id: int | None = Field(
        default=None, foreign_key="tenants.id", primary_key=True
    )
    unit_id: int | None = Field(
        default=None, foreign_key="property_units.id", primary_key=True
    )
    start_date: datetime | None = Field(
        default=None,
        sa_column=Column(DateTime(timezone=True), nullable=True)
    )
    end_date: datetime | None = Field(
        default=None,
        sa_column=Column(DateTime(timezone=True), nullable=True)
    )


class Tenant(SQLModel, table=True):
    __tablename__ = "tenants"  # type: ignore

    id: int | None = Field(default=None, primary_key=True)
    user_id: PythonUUID | None = Field(
        default=None,
        sa_column=Column(
            PG_UUID(as_uuid=True),
            ForeignKey("users.id", ondelete="SET NULL"),
            index=True,
        ),
    )
    
    # Tenant type and naming fields
    tenant_type: TenantType = Field(default=TenantType.INDIVIDUAL)
    first_name: str | None = Field(default=None, max_length=100)
    last_name: str | None = Field(default=None, max_length=100)
    company_name: str | None = Field(default=None, max_length=200)
    contact_person: str | None = Field(default=None, max_length=200)
    
    phone: str | None = None
    email: str | None = None
    status: TenantStatus = Field(default=TenantStatus.ACTIVE)
    created_at: datetime = Field(
        default_factory=create_audit_datetime,
        sa_column=Column(DateTime(timezone=True), nullable=False),
    )
    updated_at: datetime = Field(
        default_factory=create_audit_datetime,
        sa_column=Column(DateTime(timezone=True), nullable=False),
    )
    current_property_id: int | None = Field(
        default=None,
        sa_column=Column(Integer, ForeignKey(
            "properties.id", ondelete="SET NULL"))
    )
    landlord_id: PythonUUID = Field(foreign_key="users.id")
    profile_image_url: str | None = Field(
        default=None, sa_column=Column(String, nullable=True)
    )

    # QuickBooks specific fields
    quickbooks_customer_id: str | None = Field(
        default=None, sa_column=Column(String, nullable=True)
    )
    last_synced_at: datetime | None = Field(
        default=None, sa_column=Column(DateTime(timezone=True), nullable=True)
    )
    
    # Emergency contacts stored as JSONB array
    emergency_contacts: list[dict[str, Any]] = Field(
        default_factory=list,
        sa_column=Column(JSONB, nullable=True),
        description='Emergency contact information as JSONB array'
    )

    # --- Relationships Defined Directly ---

    user: Optional["User"] = Relationship(
        back_populates="tenant_details",
        sa_relationship_kwargs={"foreign_keys": "[Tenant.user_id]"},
    )
    current_property: Optional["Property"] = Relationship(
        back_populates="current_tenants")

    leases: list["Lease"] = Relationship(back_populates="tenant")
    assigned_units: list["PropertyUnit"] = Relationship(
        back_populates="tenant",
        sa_relationship_kwargs={
            "foreign_keys": "[PropertyUnit.tenant_id]",
            "lazy": "selectin"
        }
    )
    units: list["PropertyUnit"] = Relationship(
        back_populates="tenants",
        link_model=TenantUnitLink,
        sa_relationship_kwargs={"lazy": "selectin"}
    )
    maintenance_requests: list["MaintenanceRequest"] = Relationship(
        back_populates="tenant",
        sa_relationship_kwargs={"lazy": "selectin"}
    )
    payments: list["Payment"] = Relationship(
        back_populates="tenant",
        sa_relationship_kwargs={"lazy": "selectin"}
    )
    invoices: list["Invoice"] = Relationship(
        back_populates="tenant",
        sa_relationship_kwargs={"lazy": "selectin"}
    )

    @field_validator('emergency_contacts')
    @classmethod
    def validate_emergency_contacts(cls, v: list[dict[str, Any]] | None) -> list[dict[str, Any]]:
        """
        Validate and sanitize emergency contacts structure and business rules.

        Security measures implemented:
        1. XSS Prevention: HTML-escapes all text fields to prevent script injection
        2. Input Validation: Validates required fields, email format, and phone format
        3. Business Rules: Enforces maximum 5 contacts and single primary contact
        4. Database Constraints: Works in conjunction with PostgreSQL CHECK constraints

        This validator provides defense-in-depth security by:
        - Sanitizing user input at the application layer
        - Validating data integrity before database insertion
        - Preventing malicious payloads from being stored
        - Ensuring consistent data structure across the application

        Args:
            v: List of emergency contact dictionaries or None

        Returns:
            Sanitized and validated list of emergency contacts

        Raises:
            ValueError: If validation fails (max contacts, required fields, primary count, formats)

        Note: Database-level validation in migration file provides additional security layer.
        """
        if not v:
            return []

        # Security: Limit to maximum 5 contacts to prevent abuse
        if len(v) > 5:
            raise ValueError("Maximum 5 emergency contacts allowed")

        sanitized_contacts = []
        primary_count = 0

        for idx, contact in enumerate(v):
            # ============================================================================
            # REQUIRED FIELDS VALIDATION
            # ============================================================================

            # Validate name (required)
            if not contact.get('name') or not str(contact['name']).strip():
                raise ValueError(f"Emergency contact #{idx + 1}: name is required")

            # Validate phone (required)
            if not contact.get('phone') or not str(contact['phone']).strip():
                raise ValueError(f"Emergency contact #{idx + 1}: phone is required")

            # Validate relationship (required)
            if not contact.get('relationship') or not str(contact['relationship']).strip():
                raise ValueError(f"Emergency contact #{idx + 1}: relationship is required")

            # ============================================================================
            # XSS SANITIZATION: Escape HTML to prevent script injection
            # ============================================================================

            sanitized_contact = {
                # Security: HTML-escape name to prevent XSS attacks
                'name': html_escape(str(contact['name']).strip()),

                # Security: HTML-escape relationship to prevent XSS attacks
                'relationship': html_escape(str(contact['relationship']).strip()),

                # Security: Sanitize phone number (remove potential HTML/script content)
                # Keep only digits, spaces, hyphens, parentheses, and plus sign
                'phone': re.sub(r'[^\d\s\-\(\)\+]', '', str(contact['phone']).strip()),

                # Preserve boolean flag for primary contact
                'is_primary': bool(contact.get('is_primary', False)),
            }

            # ============================================================================
            # OPTIONAL FIELDS VALIDATION & SANITIZATION
            # ============================================================================

            # Email (optional) - validate format if provided
            if contact.get('email'):
                email = str(contact['email']).strip()
                if email:
                    # Basic email validation using regex
                    email_pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
                    if not re.match(email_pattern, email):
                        raise ValueError(
                            f"Emergency contact #{idx + 1}: invalid email format"
                        )
                    # Security: HTML-escape email (though emails shouldn't contain HTML)
                    sanitized_contact['email'] = html_escape(email)

            # Notes (optional) - sanitize if provided
            if contact.get('notes'):
                notes = str(contact['notes']).strip()
                if notes:
                    # Security: HTML-escape notes to prevent XSS attacks
                    # Notes are user-generated content and high-risk for XSS
                    sanitized_contact['notes'] = html_escape(notes)

            # Preserve ID if it exists (for updates), otherwise generate new UUID
            if contact.get('id'):
                sanitized_contact['id'] = contact['id']
            else:
                # Generate UUID for new contacts
                sanitized_contact['id'] = str(uuid4())

            # ============================================================================
            # BUSINESS RULE VALIDATION
            # ============================================================================

            # Count primary contacts for validation
            if sanitized_contact['is_primary']:
                primary_count += 1

            sanitized_contacts.append(sanitized_contact)

        # Ensure only one primary contact (business rule)
        if primary_count > 1:
            raise ValueError("Only one emergency contact can be marked as primary")

        return sanitized_contacts
    
    __table_args__ = (
        Index('idx_tenant_email_unique_per_landlord', "landlord_id", text("lower(email)"), unique=True, postgresql_where=text("email IS NOT NULL")),
        Index("ix_tenants_status", "status"),
        Index("idx_tenants_email_lower", text("lower(email)"),
              postgresql_where=text("email IS NOT NULL")),
        Index("ix_tenants_current_property_id", "current_property_id"),
        Index("idx_tenants_landlord_id", "landlord_id"),
        Index("ix_tenants_tenant_type", "tenant_type"),
    )

File: Frontend/src/types/tenant.ts
/**
 * Tenant type definitions with strict type safety and security considerations
 *
 * Security Note: All string fields in these interfaces should be validated
 * and sanitized before use. The backend applies XSS protection via HTML escaping.
 */

export enum TenantType {
  INDIVIDUAL = 'Individual',
  COMPANY = 'Company'
}

/**
 * Branded type for validated email addresses
 * Use this to ensure emails have passed validation before assignment
 */
export type ValidatedEmail = string & { readonly __brand: 'ValidatedEmail' };

/**
 * Branded type for validated phone numbers
 * Use this to ensure phone numbers have passed validation before assignment
 */
export type ValidatedPhone = string & { readonly __brand: 'ValidatedPhone' };

/**
 * Emergency Contact Interface
 *
 * Security Considerations:
 * - All text fields are sanitized on the backend using HTML escaping
 * - Phone numbers are stripped of non-numeric characters (except +, -, (), spaces)
 * - Email addresses must match a valid email pattern
 * - Maximum 5 emergency contacts allowed per tenant (enforced at DB level)
 * - Only one contact can be marked as primary (enforced at DB level)
 *
 * Validation Rules:
 * - name: Required, non-empty after trimming
 * - relationship: Required, non-empty after trimming (e.g., "Spouse", "Parent", "Sibling")
 * - phone: Required, non-empty after trimming
 * - email: Optional, must be valid email format if provided
 * - is_primary: Required boolean, only one can be true across all contacts
 * - notes: Optional, user-generated content (XSS risk mitigated by backend sanitization)
 *
 * @example
 * ```typescript
 * const contact: EmergencyContact = {
 *   name: 'John Doe',
 *   relationship: 'Spouse',
 *   phone: '+1-555-123-4567',
 *   email: 'john.doe@example.com',
 *   is_primary: true,
 *   notes: 'Available 24/7'
 * };
 * ```
 */
export interface EmergencyContact {
  /**
   * Client-side UUID for tracking (optional, generated on create)
   * Server does not use this field; it's for React key props and optimistic updates
   */
  readonly id?: string;

  /**
   * Full name of the emergency contact
   * Required, validated and HTML-escaped on backend to prevent XSS
   * @minLength 1 (after trimming)
   * @maxLength 200 (reasonable limit for names)
   */
  name: string;

  /**
   * Relationship to tenant (e.g., "Spouse", "Parent", "Sibling", "Friend")
   * Required, validated and HTML-escaped on backend to prevent XSS
   * @minLength 1 (after trimming)
   * @maxLength 100 (reasonable limit for relationship description)
   */
  relationship: string;

  /**
   * Contact phone number
   * Required, sanitized on backend (only digits, +, -, (), spaces allowed)
   * @pattern Must contain at least some digits
   * @example '+1-555-123-4567' or '(555) 123-4567'
   */
  phone: string;

  /**
   * Contact email address (optional)
   * Validated for proper email format if provided
   * HTML-escaped on backend (though emails shouldn't contain HTML)
   * @format email
   * @pattern ^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$
   */
  email?: string;

  /**
   * Whether this is the primary emergency contact
   * Only one contact per tenant can have this set to true (enforced at DB level)
   * @default false
   */
  is_primary: boolean;

  /**
   * Additional notes about the contact (optional)
   * User-generated content - high XSS risk, sanitized with HTML escaping on backend
   * @maxLength 500 (reasonable limit for notes)
   */
  notes?: string;
}

/**
 * Type guard to check if an emergency contact is valid
 * Use this before sending to API to catch client-side validation issues early
 *
 * @param contact - The contact object to validate
 * @returns True if contact has all required fields populated
 */
export function isValidEmergencyContact(contact: Partial<EmergencyContact>): contact is EmergencyContact {
  return (
    typeof contact.name === 'string' &&
    contact.name.trim().length > 0 &&
    typeof contact.relationship === 'string' &&
    contact.relationship.trim().length > 0 &&
    typeof contact.phone === 'string' &&
    contact.phone.trim().length > 0 &&
    typeof contact.is_primary === 'boolean'
  );
}

export enum TenantStatus {
  ACTIVE = 'Active',
  INACTIVE = 'Inactive',
  PENDING = 'Pending',
  EVICTED = 'Evicted',
  MOVED_OUT = 'Moved Out'
}

export enum MaintenancePriority {
  LOW = 'Low',
  MEDIUM = 'Medium',
  HIGH = 'High',
  URGENT = 'Urgent'
}

export enum MaintenanceStatus {
  NEW = 'New',
  PENDING = 'Pending',
  IN_PROGRESS = 'In Progress',
  SCHEDULED = 'Scheduled',
  COMPLETED = 'Completed',
  CANCELLED = 'Cancelled'
}

export enum LeaseStatus {
  DRAFT = 'DRAFT',
  PENDING = 'PENDING',
  ACTIVE = 'ACTIVE',
  EXPIRED = 'EXPIRED',
  TERMINATED = 'TERMINATED',
  RENEWED = 'RENEWED'
}

// Core Tenant Interface
export interface Tenant {
  id: number;
  user_id?: string;
  tenant_type: TenantType;
  first_name?: string;
  last_name?: string;
  company_name?: string;
  contact_person?: string;
  phone?: string;
  email?: string;
  status: TenantStatus;
  created_at: string;
  updated_at: string;
  current_property_id?: number;
  landlord_id: string;
  profile_image_url?: string;
  quickbooks_customer_id?: string;
  last_synced_at?: string;
  emergency_contacts?: EmergencyContact[];
}

// Property Unit Interface
export interface PropertyUnit {
  id: number;
  name: string;
  property_id: number;
  floor?: number;
  description?: string;
  size?: number;
  monthly_rent?: number;
  is_rented?: boolean;
  bedrooms?: number;
  bathrooms?: number;
  tenant_id?: number;
  created_at?: string;
  updated_at?: string;
  property?: {
    id: number;
    name: string;
    address?: string;
    city?: string;
    province?: string;
  };
}

// Property Interface (simplified for tenant context)
export interface Property {
  id: number;
  name: string;
  address?: string;
  city?: string;
  province?: string;
  postal_code?: string;
  property_type?: string;
  status?: string;
}

// Lease Interface
export interface Lease {
  id: number;
  start_date: string;
  end_date: string;
  monthly_rent: number;
  security_deposit: number;
  status: LeaseStatus;
  is_renewable: boolean;
  auto_renew: boolean;
  rent_due_day: number;
  late_fee_amount?: number;
  late_fee_after_days?: number;
  special_terms?: string;
  property_id: number;
  unit_id?: number;
  tenant_id: number;
  created_at: string;
  updated_at: string;
  property?: Property;
  unit?: PropertyUnit;
  documents?: LeaseDocument[];
}

// Lease Document Interface
export interface LeaseDocument {
  id: number;
  name: string;
  file_path: string;
  document_type: string;
  upload_date: string;
  lease_id: number;
  file_size?: number;
  mime_type?: string;
}

// Vendor Contact Info (basic info embedded in maintenance requests)
export interface VendorContactInfo {
  id: number;
  company_name: string;
  contact_person: string | null;
  trade_category: string;
  phone: string;
  email: string | null;
}

// Maintenance Request Interface
export interface MaintenanceRequest {
  id: number;
  issue_title: string;
  description?: string;
  property_id: number;
  unit_id?: number;
  tenant_id?: number;
  user_id?: string;
  request_date: string;
  priority: MaintenancePriority;
  status: MaintenanceStatus;
  scheduled_date?: string;
  completed_date?: string;
  estimated_cost?: number;
  actual_cost?: number;
  photos?: string[];
  created_at: string;
  updated_at: string;
  assigned_to?: string;
  vendor_id?: number | null;
  vendor?: VendorContactInfo | null;
  notify_tenant: boolean;
  property?: Property;
  unit?: PropertyUnit;
  tenant?: Tenant;  // Populated by backend when loading maintenance requests
  preferred_time?: string;
}

// Payment Interface
export interface Payment {
  id: number;
  amount: number;
  payment_date: string;
  payment_method?: string;
  transaction_reference?: string;
  description?: string;
  status: 'Pending' | 'Paid' | 'Partial' | 'Overdue' | 'Cancelled' | 'Refunded' | 'Draft' | 'Void' | 'Uncollectible';
  tenant_id: number;
  lease_id?: number;
  reduction_amount?: number;
  reduction_reason?: string;
  receipt_url?: string;
  created_at: string;
  updated_at: string;
}

// Invoice Interface
export interface Invoice {
  id: number;
  invoice_number: string;
  amount: number;
  due_date: string;
  issue_date: string;
  description: string;
  status: 'Pending' | 'Paid' | 'Partial' | 'Overdue' | 'Cancelled' | 'Refunded' | 'Draft' | 'Void' | 'Uncollectible';
  tenant_id: number;
  property_id?: number;
  created_at: string;
  updated_at: string;
}

// Enriched Tenant with all relationships
export interface EnrichedTenant extends Tenant {
  leases?: Lease[];
  maintenance_requests?: MaintenanceRequest[];
  documents?: LeaseDocument[];
  payments?: Payment[];
  invoices?: Invoice[];
  unit?: PropertyUnit;
  property?: Property;
  assigned_units?: PropertyUnit[];
  units?: PropertyUnit[];
}

// API Response Types
export interface TenantResponse extends EnrichedTenant {
  // Additional computed fields that might come from backend
  full_name?: string;
  display_name?: string;
}

// Component Prop Types
export interface TenantTableProps {
  tenants: EnrichedTenant[];
  onAddTenant: () => void;
  isLoading: boolean;
  selectedTenants: number[];
  onToggleSelectAll: () => void;
  onToggleSelect: (tenantId: number) => void;
}

// Maintenance Form Types
export interface MaintenanceFormData {
  issue_title: string;
  description?: string;
  priority: MaintenancePriority | 'Low' | 'Medium' | 'High';
  status: MaintenanceStatus | 'Pending' | 'In Progress' | 'Scheduled' | 'Completed' | 'Cancelled';
  property_id: string;
  unit_id?: string;
  tenant_id?: string;
  assigned_to?: string;
  vendor_id?: string;
  notify_tenant?: boolean;
  scheduled_date?: string;
  estimated_cost?: string;
  start_date?: string;
  end_date?: string;
  photos?: string[];
  preferred_time?: string;
}

// Photo Upload Types
export interface PhotoFileWithId {
  id: string;
  file: File;
  name: string;
  size: number;
  preview?: string; // Object URL for local preview before upload
}

export interface PhotoUploadProgress {
  id: string;
  status: 'pending' | 'uploading' | 'done' | 'error';
  error?: string;
}

export interface MaintenancePhotoState {
  selectedFiles: PhotoFileWithId[];
  uploadingPhotos: boolean;
  uploadProgress: PhotoUploadProgress[];
  uploadError: string | null;
}

// Maintenance Summary Type
export interface MaintenanceSummary {
  total_requests: number;
  new: number;
  pending: number;
  in_progress: number;
  completed: number;
  cancelled: number;
}

// Table Props
export interface MaintenanceTableProps {
  requests: MaintenanceRequest[];
  onEdit: (request: MaintenanceRequest) => void;
  onDelete: (requestId: number) => void;
  onView: (request: MaintenanceRequest) => void;
  currentPage?: number;
  pageSize?: number;
}